File: 1606.find-servers-that-handled-most-number-of-requests.cpp
/*
 * @lc app=leetcode id=1606 lang=cpp
 *
 * [1606] Find Servers That Handled Most Number of Requests
 */
#include <bits/stdc++.h>

using std::pair;
using std::vector;

// @lc code=start
class Solution {
public:
    // T: O(n * log(k)), n := no. of requests
    // S: O(n + k)
    vector<int> busiestServers(int k, vector<int>& arrival, vector<int>& load) {
        vector<int> srv_reqs(k, 0);

        std::set<int> avail;
        for (int i = 0; i < k; ++i) {
            avail.insert(i);
        }

        // PQ with elements (end_time, srv_id)
        std::priority_queue<pair<int, int>, vector<pair<int, int>>, greater<>> busy;
        
        for (int i = 0; i < arrival.size(); ++i) {
            int t_cur = arrival[i], t_end = t_cur + load[i];

            // Find available servers with end time of previous request
            // less than or equal to the time of current request
            while (busy.size() && busy.top().first <= t_cur) {
                // busy.pop() at most (n - k) times
                int srv = busy.top().second; busy.pop();
                avail.insert(srv);
            }

            if (avail.empty()) continue;  // all busy, drop request

            // Assign req to the first available server whose index 
            // is greater than i % k in circular order. 
            auto it = avail.lower_bound(i % k);
            if (it == avail.end())  it = avail.begin();
            ++srv_reqs[*it];
            busy.emplace(t_end, *it);
            avail.erase(*it);
        }

        // Find servers that handled the most requests
        int max_reqs = *std::max_element(srv_reqs.begin(), srv_reqs.end());
        vector<int> res;
        res.reserve(k);
        for (int i = 0; i < k; ++i) {
            if (srv_reqs[i] == max_reqs) {
                res.push_back(i);
            }
        }
        return res;
    }
};
// @lc code=end

File: 1606.find-servers-that-handled-most-number-of-requests.py
#
# @lc app=leetcode id=1606 lang=python3
#
# [1606] Find Servers That Handled Most Number of Requests
#
import heapq
from typing import List

# @lc code=start
class Solution:
    # T: O(n * log(k)), n := the number of requests
    # S: O(k)
    def busiestServers(self, k: int, arrival: List[int], load: List[int]) -> List[int]:
        avail = list(range(k))
        busy = []
        n_handled = [0] * k

        for i, (arr_t, process_t) in enumerate(zip(arrival, load)):
            while len(busy) and busy[0][0] <= arr_t:  # check if any servers finish processing prev. reqs
                _, j = heapq.heappop(busy)
                # remap j so that it's within interval [i, i + k - 1], 
                # and push it to avail server id heap
                heapq.heappush(avail, i + (j - i) % k)

            if len(avail):
                j = heapq.heappop(avail) % k
                # arr_t + process_t will be the end time of the req being handled
                heapq.heappush(busy, (arr_t + process_t, j))
                n_handled[j] += 1

            # drop req if no avail server
        
        max_handled = max(n_handled)
        return [i for i in range(k) if n_handled[i] == max_handled]

# @lc code=end

